### PR TITLE
 Enableable ArrayNodeDefinition is disabled for empty configuration

### LINKED ISSUE #25760: [DependencyInjection] ArrayNodeDefinition::canBeEnabled() doesn't make the configuration disabled by default
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 2.7

Consider the following configuration:

```php
$builder = new TreeBuilder();
$root = $builder->root('acme_foo');
$root->canBeEnabled();
```

**Expected behaviour**

If the section wasn't provided in configuration, it will be disabled.

**Actual Behaviour**

The section is enabled unless it was explicitly disabled. 

**Supporting information**

It only seems to be an issue for root nodes. Any child nodes work as expected (i.e. `$root->children()->arrayNode('foo')->canBeEnabled()`).

The docblock for `canBeEnabled()` says that the section should be disabled by default (which is also what I expected):

```php
    /**
     * Adds an "enabled" boolean to enable the current section.
     *
     * By default, the section is disabled. If any configuration is specified then
     * the node will be automatically enabled:
     *
     * enableableArrayNode: {enabled: true, ...}   # The config is enabled & default values get overridden
     * enableableArrayNode: ~                      # The config is enabled & use the default values
     * enableableArrayNode: true                   # The config is enabled & use the default values
     * enableableArrayNode: {other: value, ...}    # The config is enabled & default values get overridden
     * enableableArrayNode: {enabled: false, ...}  # The config is disabled
     * enableableArrayNode: false                  # The config is disabled
     */
```

### PR DESCRIPTION
**Pull request by mateuszsip**
| Q             | A
| ------------- | ---
| Branch?       | 2.7+
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #25760
| License       | MIT

Fixes #25760.

Currently, documented behavior is not true:

https://github.com/symfony/symfony/blob/70c8c2d47bd71861c8205985a17e68cedf828e1f/src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php#L207-L208

### CODE REVIEW

File: src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php [hunk 1]
```diff
@@ -226,7 +226,9 @@ public function canBeEnabled()
             ->beforeNormalization()
                 ->ifArray()
                 ->then(function ($v) {
-                    $v['enabled'] = isset($v['enabled']) ? $v['enabled'] : true;
+                    if (!isset($v['enabled'])) {
+                        $v['enabled'] = empty($v) ? false : true;
```
Comment (iltar) [inline src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php]: Can also be written as `$v['enabled'] = !empty($v);`

Comment (kejwmen) [reply src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php]: damn, you're right :D

**Review by Simperfit — APPROVED** [review]
The tests failure seems unrelated to the current PR.

**Review by xabbuh — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @kejwmen.

**Comment by fabpot** [comment]
reverted because of #25862 (9312f794e8)

File: src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php [hunk 2]
```diff
@@ -226,7 +226,9 @@ public function canBeEnabled()
             ->beforeNormalization()
                 ->ifArray()
                 ->then(function ($v) {
-                    $v['enabled'] = isset($v['enabled']) ? $v['enabled'] : true;
+                    if (!isset($v['enabled'])) {
+                        $v['enabled'] = !empty($v);
```
Comment (jakzal) [inline src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php]: Should probably be `is_array($v) && !empty($v)` to cover the `foo: ~` case, which should work as enable.

Comment (kejwmen) [reply src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php]: What tests should we add to make sure it won't happen in the future? Component tests cover `null` case already. `~` is just a `null`, right?

Comment (jakzal) [reply src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php]: forget my comment. This callback is only triggered for arrays ;)

Comment (kejwmen) [reply src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php]: Looks like `~` in yaml is represented by empty array, simple reproducer is available [here](https://github.com/symfony/symfony/issues/25862)

**Pull request closed without merging**